scipy/spatial/transform/rotation.py

```diff
@@ -3,12 +3,127 @@
+    n2 = elementary_basis_vector(seq[1])
+    n3 = elementary_basis_vector(seq[2])
+
+    # Step 2
+    sl = np.dot(np.cross(n1, n2), n3)
+    cl = np.dot(n1, n3)
+
+    # angle offset is lambda from the paper referenced in [2] from docstring of
+    # `as_euler` function
+    offset = np.arctan2(sl, cl)
+    c = np.empty((3, 3))
+    c[0] = n2
+    c[1] = np.cross(n1, n2)
+    c[2] = n1
+
+    # Step 3
+    rt = np.array([
+        [1, 0, 0],
+        [0, cl, -sl],
+        [0, sl, cl]
+    ])
+    rtc = rt.dot(c)
+    res = np.einsum('...ij,...jk->...ik', rtc, dcm)
+    dcm_transformed = np.einsum('...ij,...jk->...ik', res, c.T)
+
+    # Step 4
+    angles = np.empty((num_rotations, 3))
+    angles[:, 1] = np.arccos(dcm_transformed[:, 2, 2])
+
+    # Steps 5, 6
+    eps = 1e-7
+    safe1 = (np.abs(angles[:, 1]) >= eps)
+    safe2 = (np.abs(angles[:, 1] - np.pi) >= eps)
+
+    # Step 4 (Completion)
+    angles[:, 1] += offset
+
+    # 5b
+    safe_mask = np.logical_and(safe1, safe2)
+    angles[safe_mask, 0] = np.arctan2(dcm_transformed[safe_mask, 2, 0],
+                                      -dcm_transformed[safe_mask, 2, 1])
+    angles[safe_mask, 2] = np.arctan2(dcm_transformed[safe_mask, 0, 2],
+                                      dcm_transformed[safe_mask, 1, 2])
+
+    # 6a
+    angles[~safe_mask, 2] = 0
+    # 6b
+    angles[~safe1, 0] = np.arctan2(
+        dcm_transformed[~safe1, 0, 1] - dcm_transformed[~safe1, 1, 0],
+        dcm_transformed[~safe1, 0, 0] + dcm_transformed[~safe1, 1, 1]
+    )
+    # 6c
+    angles[~safe2, 0] = np.arctan2(
+        dcm_transformed[~safe2, 0, 1] + dcm_transformed[~safe2, 1, 0],
+        dcm_transformed[~safe2, 0, 0] - dcm_transformed[~safe2, 1, 1]
+    )
+
+    # Step 7
+    # python modulo operator works correctly for negative numbers
+    if seq[0] == seq[2]:
+        # lambda = 0, so we can only ensure angle2 -> [0, pi]
+        adjust_mask = np.logical_or(angles[:, 1] < 0, angles[:, 1] > np.pi)
+    else:
+        # lambda = + or - pi/2, so we can ensure angle2 -> [-pi/2, pi/2]
+        adjust_mask = np.logical_or(angles[:, 1] < -np.pi / 2,
+                                    angles[:, 1] > np.pi / 2)
+
+    # Dont adjust gimbal locked angle sequences
+    adjust_mask = np.logical_and(adjust_mask, safe_mask)
+
+    angles[adjust_mask, 0] += np.pi
+    angles[adjust_mask, 1] = 2 * offset - angles[adjust_mask, 1]
+    angles[adjust_mask, 2] -= np.pi
+
+    angles[angles < -np.pi] += 2 * np.pi
+    angles[angles > np.pi] -= 2 * np.pi
+
+    # Step 8
+    # TODO: if any observability flags are poor, possibly raise a UserWarning?
+    if not np.all(safe_mask):
+        warnings.warn("Some angle sequences suffer from gimbal lock. In those "
+                      "cases, it is not possible to determine the angles "
+                      "uniquely.", UserWarning, stacklevel=1)
+
+    # Reverse role of extrinsic and intrinsic rotations
+    return angles[:, ::-1] if extrinsic else angles
```

Comment:
In case of gimbal lock we will have for example [10, 90, 0] and then after reversion [0, 90, 10]. My idea/suggestion is to perhaps always return the third angle as zero for intrinsic and extrinsic sequences.

I think you can simply swap angles 1 and 3 for gimbal lock cases in extrinsic sequences. What do you think?

Reply:
Looks like you can write 
```
if extrinsic:
   angles[safe_mask] = angles[safe_mask, ::-1]
```

Reply:
So doing this causes the tests to fail. I inspected the dcm of the returned and it looks like not reversing the angle sequences leads to sign reversals of `dcm[:, 0, 1]` and `dcm[:, 1, 2]`. Both of these quantities are small and close to zero, but the change in sign leads to assertion failure.

What do you suggest? Should we say that the first angle is zero for extrinsic rotations?